src/swaps/status_stream.rs

```diff
@@ -0,0 +1,183 @@
+            return Ok(());
+        }
+        self.send_json(&WsRequest::subscribe_swaps_request(ids))
+            .await
+    }
+
+    async fn send_json(&mut self, data: &WsRequest) -> Result<(), Error> {
+        let t = serde_json::to_string(data)?;
+        self.ws.send(Message::text(t)).await?;
+        Ok(())
+    }
+}
+
+pub struct BoltzWsApi {
+    ws_url: String,
+    swap_ids: Mutex<Vec<String>>,
+    subscription_notifier: broadcast::Sender<String>,
+    update_notifier: broadcast::Sender<SwapStatus>,
+    shutdown_sender: Mutex<Option<oneshot::Sender<()>>>,
+}
+
+impl BoltzWsApi {
+    pub fn new(ws_url: String) -> Self {
+        let (subscription_notifier, _) = broadcast::channel(16);
+        let (update_notifier, _) = broadcast::channel(16);
+        Self {
+            ws_url,
+            swap_ids: Mutex::new(vec![]),
+            subscription_notifier,
+            update_notifier,
+            shutdown_sender: Mutex::new(None),
+        }
+    }
+
+    pub fn updates(&self) -> broadcast::Receiver<SwapStatus> {
+        self.update_notifier.subscribe()
+    }
+
+    pub fn subscribe(&self, swap_id: &str) -> Result<(), Error> {
+        let _ = self.subscription_notifier.subscribe();
+        match self.subscription_notifier.send(swap_id.to_string()) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(Error::Protocol(format!(
+                "Failed to send subscription: {:?}",
+                e
+            ))),
+        }
+    }
+
+    pub fn start(self: Arc<Self>) {
+        let keep_alive_ping_interval = Duration::from_secs(15);
+        let reconnect_delay = Duration::from_secs(2);
+        let mut sub_stream = self.subscription_notifier.subscribe();
+
+        tokio::spawn(async move {
+            let (shutdown_sender, mut shutdown_receiver) = oneshot::channel();
+            let _ = self.shutdown_sender.lock().await.replace(shutdown_sender);
+
+            loop {
+                match BoltzWsConnection::new(self.ws_url.as_str()).await {
+                    Ok(mut connection) => {
+                        {
+                            let ids = self.swap_ids.lock().await;
+                            match connection.subscribe(ids.to_owned()).await {
+                                Ok(_) => {}
+                                Err(e) => {
+                                    error!("Error subscribing to swaps: {:?}", e);
+                                    tokio::time::sleep(reconnect_delay).await;
+                                    continue;
+                                }
+                            }
+                        }
+                        let mut interval = tokio::time::interval(keep_alive_ping_interval);
+
+                        loop {
+                            tokio::select! {
+                                _ = &mut shutdown_receiver => {
+                                    info!("Received shutdown signal, exiting socket loop");
+                                    break;
```

Comment:
If I am not mistaken this will only break the inner loop, and cause a reconnection. You can use labels to break the outer loop from this line